clippy_lints/src/methods/filter_map.rs

```diff
@@ -50,6 +52,197 @@ fn is_option_filter_map(cx: &LateContext<'_>, filter_arg: &hir::Expr<'_>, map_ar
+                            OffendingFilterExpr::IsSome { .. } => CalledMethod::OptionIsSome,
+                            OffendingFilterExpr::IsOk { .. } => CalledMethod::ResultIsOk,
+                            OffendingFilterExpr::Matches { .. } => unreachable!("only IsSome and IsOk can get here"),
+                        }
+                    })
+                } else {
+                    None
+                }
+            },
+            OffendingFilterExpr::Matches { variant_def_id } => {
+                let expr_uses_local = |pat: &Pat<'_>, expr: &Expr<'_>| {
+                    if let PatKind::TupleStruct(QPath::Resolved(_, path), [subpat], _) = pat.kind
+                        && let PatKind::Binding(_, local_id, ident, _) = subpat.kind
+                        && path_to_local_id(expr.peel_blocks(), local_id)
+                        && let Some(local_variant_def_id) = path.res.opt_def_id()
+                        && local_variant_def_id == variant_def_id
+                    {
+                        Some((ident, pat.span))
+                    } else {
+                        None
+                    }
+                };
+
+                // look for:
+                // `if let Variant   (v) =         enum { v } else { unreachable!() }`
+                //         ^^^^^^^    ^            ^^^^            ^^^^^^^^^^^^^^^^^^
+                //    variant_span  variant_ident  scrutinee       else_ (blocks peeled later)
+                // OR
+                // `match enum {   Variant       (v) => v,      _ => unreachable!() }`
+                //        ^^^^     ^^^^^^^        ^                  ^^^^^^^^^^^^^^
+                //     scrutinee  variant_span  variant_ident        else_
+                let (scrutinee, else_, variant_ident, variant_span) =
+                    match higher::IfLetOrMatch::parse(cx, map_body.value) {
+                    Some(higher::IfLetOrMatch::IfLet(sc, pat, then, Some(else_)))
+                        if let Some((ident, span)) = expr_uses_local(pat, then) => {
+                        (sc, else_, ident, span)
+                    },
+                    Some(higher::IfLetOrMatch::Match(sc, [arm1, arm2], MatchSource::Normal)) => {
+                        let is_variant_arm = |arm: &hir::Arm<'_>, other_arm: &hir::Arm<'_>| {
+                            // check if the other arm has pattern `_` and this arm references the local (modulo blocks)
+                            if let PatKind::Wild = other_arm.pat.kind {
+                                expr_uses_local(arm.pat, arm.body.peel_blocks())
+                            } else {
+                                None
+                            }
+                        };
+                        let (else_arm, (ident, span)) = is_variant_arm(arm1, arm2)
+                            .map(|data| (arm2, data))
+                            .or_else(|| is_variant_arm(arm2, arm1).map(|data| (arm1, data)))?;
```

Comment:
We don't need to test both orientations, the `_` arm must be the second one because

```rust
match x {
    _ => {},
//  ^ this arm always matches
    None => {},
//  ^^^^ unreachable
}
```